### DISCUSSION TITLE
How can I hide the aggreagate audio device?

### DISCUSSION THREAD

**Discussion by dttlgotv** [ExistentialAudio/BlackHole#752] [Q&A]
I am macox developer. I can use coreaudio api to create a aggregated device. In fact it is multi-output device including Blackhole and system speaker.

I can not find  api to hide the aggregate device.  The created aggregated device is always showed in Midi.

**Comment by gchilds** [comment]
is setting `kAudioAggregateDeviceIsPrivateKey` to `1` not working for you?

**Reply by dttlgotv** [reply]
Not working

**Reply by gchilds** [reply]
can you share the aggregate setup code @dttlgotv?

**Reply by dttlgotv** [reply]
//-----------------------
    // Create a CFDictionary for our aggregate device
    //-----------------------

    CFMutableDictionaryRef aggDeviceDict = CFDictionaryCreateMutable(NULL, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);

    CFStringRef AggregateDeviceNameRef = CFSTR(EEO_AGG_OUT_NAME);
    CFStringRef AggregateDeviceUIDRef = CFSTR(EEO_AGG_OUT_UID);

    // add the name of the device to the dictionary
    CFDictionaryAddValue(aggDeviceDict, CFSTR(kAudioAggregateDeviceNameKey), AggregateDeviceNameRef);

    // add our choice of UID for the aggregate device to the dictionary
    CFDictionaryAddValue(aggDeviceDict, CFSTR(kAudioAggregateDeviceUIDKey), AggregateDeviceUIDRef);

    // make it private so that we don't have the user messing with it
    int value = 1;
    CFDictionaryAddValue(aggDeviceDict, CFSTR(kAudioAggregateDeviceIsPrivateKey), CFNumberCreate(NULL, kCFNumberIntType, &value));


    int private_value = 0;
    CFNumberRef aggregate_device_private_key = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &private_value);
    CFDictionaryAddValue(aggDeviceDict, CFSTR(kAudioAggregateDeviceIsPrivateKey), aggregate_device_private_key);
    CFRelease(aggregate_device_private_key);

    int stacked_value = 1;
    CFNumberRef aggregate_device_stacked_key = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &stacked_value);
    CFDictionaryAddValue(aggDeviceDict, CFSTR(kAudioAggregateDeviceIsStackedKey), aggregate_device_stacked_key);
    CFRelease(aggregate_device_stacked_key);

    //-----------------------
    // Feed the dictionary to the plugin, to create a blank aggregate device
    //-----------------------

**Reply by dttlgotv** [reply]
After I set kAudioAggregateDeviceIsPrivateKey to 1, I can see the aggreagate device in Midi too.

**Reply by gchilds** [reply]
Your dictionary looks ok, and my attempts to create a private stacked aggregate device work, but why do you set `kAudioAggregateDeviceIsPrivateKey` twice?

Can you pick one method? And show the actual aggregate creation?

**Reply by dttlgotv** [reply]
//-----------------------
    // Create a CFDictionary for our aggregate device
    //-----------------------

    CFMutableDictionaryRef aggDeviceDict = CFDictionaryCreateMutable(NULL, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);

    CFStringRef AggregateDeviceNameRef = CFSTR(EEO_AGG_OUT_NAME);
    CFStringRef AggregateDeviceUIDRef = CFSTR(EEO_AGG_OUT_UID);

    // add the name of the device to the dictionary
    CFDictionaryAddValue(aggDeviceDict, CFSTR(kAudioAggregateDeviceNameKey), AggregateDeviceNameRef);

    // add our choice of UID for the aggregate device to the dictionary
    CFDictionaryAddValue(aggDeviceDict, CFSTR(kAudioAggregateDeviceUIDKey), AggregateDeviceUIDRef);


    int private_value = 0;
    CFNumberRef aggregate_device_private_key = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &private_value);
    CFDictionaryAddValue(aggDeviceDict, CFSTR(kAudioAggregateDeviceIsPrivateKey), aggregate_device_private_key);
    CFRelease(aggregate_device_private_key);

    int stacked_value = 1;
    CFNumberRef aggregate_device_stacked_key = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &stacked_value);
    CFDictionaryAddValue(aggDeviceDict, CFSTR(kAudioAggregateDeviceIsStackedKey), aggregate_device_stacked_key);
    CFRelease(aggregate_device_stacked_key);

    //-----------------------
    // Feed the dictionary to the plugin, to create a blank aggregate device
    //-----------------------

    AudioObjectPropertyAddress pluginAOPA;
    pluginAOPA.mSelector = kAudioPlugInCreateAggregateDevice;
    pluginAOPA.mScope = kAudioObjectPropertyScopeGlobal;
    pluginAOPA.mElement = kAudioObjectPropertyElementMaster;//kAudioObjectPropertyScopeOutput;
    UInt32 outDataSize;

    osErr = AudioObjectGetPropertyDataSize(pluginID, &pluginAOPA, 0, NULL, &outDataSize);
    if (osErr != noErr) {
        qWarning("audio:err: CreateAggregateDevice line is %d err %d", __LINE__,  osErr);
        return osErr;
    }


    osErr = AudioObjectGetPropertyData(pluginID, &pluginAOPA, sizeof(aggDeviceDict), &aggDeviceDict, &outDataSize, &outAggregateDevice);
    if (osErr != noErr) {
        qWarning("audio:err: CreateAggregateDevice line is %d err %d", __LINE__,  osErr);
        return osErr;
    }


    // pause for a bit to make sure that everything completed correctly
    // this is to work around a bug in the HAL where a new aggregate device seems to disappear briefly after it is created
    CFRunLoopRunInMode(kCFRunLoopDefaultMode, 0.1, false);


    //-----------------------
    // Create a CFMutableArray for our sub-device list
    //-----------------------

    // this example assumes that you already know the UID of the device to be added
    // you can find this for a given AudioDeviceID via AudioDeviceGetProperty for the kAudioDevicePropertyDeviceUID property
    // obviously the example deviceUID below won't actually work!
    AudioDeviceID devId;
    devId = getDefaultOutDeviceId();
    qInfo() << "audio:agg get default out device id is "<<devId;
    if(0==devId){
        qWarning() << "audio:err: fail to get default out device id";
        return -1;
    }

   
    CFStringRef defaultUID = GetDeviceUID(devId);
    if(0==defaultUID){
        qWarning() << "audio:err: fail to get default GetDeviceUID";
        return -1;
    }



    CFStringRef  eeoUID = GetDeviceUID(devId);
    if(0==eeoUID){
        qWarning() << "audio:err: fail to get eeo GetDeviceUID";
        return -1;
    }

    // we need to append the UID for each device to a CFMutableArray, so create one here
    CFMutableArrayRef subDevicesArray = CFArrayCreateMutable(NULL, 0, &kCFTypeArrayCallBacks);

    // just the one sub-device in this example, so append the sub-device's UID to the CFArray
    CFArrayAppendValue(subDevicesArray, defaultUID);
    CFArrayAppendValue(subDevicesArray, eeoUID);

    // if you need to add more than one sub-device, then keep calling CFArrayAppendValue here for the other sub-device UIDs

    //-----------------------
    // Set the sub-device list
    //-----------------------

    pluginAOPA.mSelector = kAudioAggregateDevicePropertyFullSubDeviceList;
    pluginAOPA.mScope = kAudioObjectPropertyScopeGlobal;
    pluginAOPA.mElement = kAudioObjectPropertyElementMaster;//kAudioObjectPropertyScopeOutput;
    outDataSize = sizeof(CFMutableArrayRef);
    osErr = AudioObjectSetPropertyData(outAggregateDevice, &pluginAOPA, 0, NULL, outDataSize, &subDevicesArray);
    if (osErr != noErr) {
        qWarning("audio:err: CreateAggregateDevice line is %d err %d", __LINE__,  osErr);
        return osErr;
    }

    
    // pause again to give the changes time to take effect
    CFRunLoopRunInMode(kCFRunLoopDefaultMode, 0.1, false);

**Reply by gchilds** [reply]
Huh, interesting. I didn't know you could modify a "blank" aggregate device. So _that's_ how `Audio MIDI Setup` does it. 
I guess your several step setup is blowing away the private flag. Thing is you don't need to create the aggregate device in several steps because you can create it in one with a single dictionary and call to `AudioHardwareCreateAggregateDevice()`.

Try adding the keys `kAudioAggregateDeviceSubDeviceListKey`, `kAudioAggregateDeviceMasterSubDeviceKey` to your dictionary and calling `AudioHardwareCreateAggregateDevice()` instead of doing all that plug in stuff and waiting for core audio.

**Reply by dttlgotv** [reply]
It works now .Thanks a lot

**Reply by gchilds** [reply]
You're welcome. That must have been very old code.

**Reply by dttlgotv** [reply]
It is very suprised. After I hide the aggreagate device,  when I use this aggreagate device to playing something, everything is silence. And when I show the aggreagate device, everything is well.

**Reply by gchilds** [reply]
How are you playing audio to the aggregate device? 
By making it the default output device? I don't think a private aggregate can be default output.

**Reply by dttlgotv** [reply]
Yes. I just set it to default output for recording system audio.

**Reply by gchilds** [reply]
That won’t work, also aggregates don’t make a good default output because they have no volume control. You should probably make a loop back device (like black hole) be the default output and then forward its input to your stacked, private aggregate, but this is getting a little weird, so


what are you trying to do?

**Reply by dttlgotv** [reply]
I have to use aggreate (blackhole+system default ouput) as default output for recording system audio. It is a little bad as you known that the aggregate device can not support volume control.  If I only use blackhole as default audio system, currently no audio can be heard because blackhole only loopback the audio and it can not support output the audio to real speaker.

I think the best idea is making the blackhole output the audio to real speakek, then the audio can be heard and system audio can be recorded too.

any idea about how to enhance the blackhole ？ 
Thanks a lot.

**Comment by dttlgotv** [comment]
> kAudioAggregateDeviceIsPrivateKey



> Your dictionary looks ok, and my attempts to create a private stacked aggregate device work, but why do you set `kAudioAggregateDeviceIsPrivateKey` twice?
> 
> Can you pick one method? And show the actual aggregate creation?
